pkg/factory/factory.go

```diff
@@ -177,37 +177,41 @@ func (rf *resourceFactory) GetDeviceProvider(dt types.DeviceType) types.DevicePr
 	}
 }
 
-// GetDeviceFilter unmarshal the "selector" values from ResourceConfig and returns an instance of DeviceSelector based on
-// DeviceType in the ResourceConfig
-func (rf *resourceFactory) GetDeviceFilter(rc *types.ResourceConfig) (interface{}, error) {
-	switch rc.DeviceType {
-	case types.NetDeviceType:
-		netDeviceSelector := &types.NetDeviceSelectors{}
-
-		if err := json.Unmarshal(*rc.Selectors, netDeviceSelector); err != nil {
-			return nil, fmt.Errorf("error unmarshalling NetDevice selector bytes %v", err)
+// parseObjectOrSlice unmarshal's the "Selector" values from the ResourceConfig into a slice of *DeviceSelectors.
+// Each *DeviceSelector has been converted to any before being returned. parseObjectOrSlice will parse both
+// kinds of valid "selector" values - a slice or a single object.
+func parseObjectOrSlice[O types.NetDeviceSelectors | types.AccelDeviceSelectors | types.AuxNetDeviceSelectors](
+	rc *types.ResourceConfig) ([]any, error) {
```

Comment:
you can return here ([]*O, error) imo and it will still work in GetDeviceFilter below.

that way its a bit more explicit on what this does.

Reply:
I tried this change, and it did not compile. The error was: `cannot use interfaceArray (variable of type []any) as type []*O in return statement`. I could do the interface object conversion outside of this function, but I think it makes more sense to have that conversion logic here rather than in `GetDeviceFilter`.

Unless I misunderstood your suggestion, I think the return type should to remain `([]any, error)`.

Reply:
yes ! you are right, lets keep it as is.
(tried it myself, current impl makes more  sense)